test-suite/src/test/scala/org/scalajs/testsuite/scalalib/ArrayBuilderTest.scala

```diff
@@ -0,0 +1,266 @@
+      expect(erase(a(1)).equals('S')).toBeTruthy
+    }
+
+    it("Char, zeros inline") {
+      val a = zerosInline[Char](3)
+      expect(a.getClass).toBe(classOf[Array[Char]])
+      expect(a.length).toEqual(3)
+      expect(erase(a(0)).isInstanceOf[Char]).toBeTruthy
+      expect(erase(a(0)).equals('\0')).toBeTruthy
+    }
+
+    it("Char, zeros noinline") {
+      val a = zerosNoInline[Char](3)
+      expect(a.getClass).toBe(classOf[Array[Char]])
+      expect(a.length).toEqual(3)
+      expect(erase(a(0)).isInstanceOf[Char]).toBeTruthy
+      expect(erase(a(0)).equals('\0')).toBeTruthy
+    }
+
+    it("Boolean, normal case inline") {
+      val b = ArrayBuilder.make[Boolean]()
+      b += true
+      b += someBoolean
+      val a = b.result()
+
+      expect(a.getClass).toBe(classOf[Array[Boolean]])
+      expect(a.length).toEqual(2)
+      expect(erase(a(0)).isInstanceOf[Boolean]).toBeTruthy
+      expect(erase(a(0)).equals(true)).toBeTruthy
+      expect(erase(a(1)).equals(false)).toBeTruthy
+    }
+
+    it("Boolean, normal case noinline") {
+      val b = makeNoInline[Boolean]()
+      b += true
+      b += someBoolean
+      val a = b.result()
+
+      expect(a.getClass).toBe(classOf[Array[Boolean]])
+      expect(a.length).toEqual(2)
+      expect(erase(a(0)).isInstanceOf[Boolean]).toBeTruthy
+      expect(erase(a(0)).equals(true)).toBeTruthy
+      expect(erase(a(1)).equals(false)).toBeTruthy
+    }
+
+    it("Boolean, zeros inline") {
+      val a = zerosInline[Boolean](3)
+      expect(a.getClass).toBe(classOf[Array[Boolean]])
+      expect(a.length).toEqual(3)
+      expect(erase(a(0)).isInstanceOf[Boolean]).toBeTruthy
+      expect(erase(a(0)) == false).toBeTruthy
+    }
+
+    it("Boolean, zeros noinline") {
+      val a = zerosNoInline[Boolean](3)
+      expect(a.getClass).toBe(classOf[Array[Boolean]])
+      expect(a.length).toEqual(3)
+      expect(erase(a(0)).isInstanceOf[Boolean]).toBeTruthy
+      expect(erase(a(0)) == false).toBeTruthy
+    }
+
+    it("Unit, normal case inline") {
+      val b = ArrayBuilder.make[Unit]()
+      b += (())
+      val a = b.result()
+
+      expect(a.getClass).toBe(classOf[Array[Unit]])
+      expect(a.length).toEqual(1)
+      expect(erase(a(0)).isInstanceOf[Unit]).toBeTruthy
+      expect(erase(a(0)).equals(())).toBeTruthy
+    }
+
+    it("Unit, normal case noinline") {
+      val b = makeNoInline[Unit]()
+      b += (())
+      val a = b.result()
+
+      expect(a.getClass).toBe(classOf[Array[Unit]])
+      expect(a.length).toEqual(1)
+      expect(erase(a(0)).isInstanceOf[Unit]).toBeTruthy
+      expect(erase(a(0)).equals(())).toBeTruthy
+    }
+
+    it("Unit, zeros inline") {
+      val a = zerosInline[Unit](3)
+      expect(a.getClass).toBe(classOf[Array[Unit]])
+      expect(a.length).toEqual(3)
+      expect(erase(a(0)).isInstanceOf[Unit]).toBeTruthy // JVM says false
+      expect(erase(a(0)) == (())).toBeTruthy            // JVM says false
```

Comment:
No, it's `()`. When you unbox `null` to `Unit` you get `()`, the same way that if you unbox `null` to an `Int` you get 0. `null` is the zero of `BoxedUnit`. The JVM ends up with `null` here, but that is an accident of how they wrote `ArrayBuilder.ofUnit`; they bypass the unboxing that should fix `null` into `()` (I think I'll file a bug report about that, the fix is simple.)